### ISSUE TITLE
API Proposal: File.ReadLinesAsync

### ISSUE THREAD

**Issue by khellang** [dotnet/runtime#2214]
I'd like to propose the following new methods to the `File` type:

```diff
namespace System.IO {
    public static partial class File {
        public static System.Collections.Generic.IEnumerable<string> ReadLines(string path) { throw null; }
        public static System.Collections.Generic.IEnumerable<string> ReadLines(string path, System.Text.Encoding encoding) { throw null; }
+       public static System.Collections.Generic.IAsyncEnumerable<string> ReadLinesAsync(string path, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
+       public static System.Collections.Generic.IAsyncEnumerable<string> ReadLinesAsync(string path, System.Text.Encoding encoding, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
    }
}
```

The methods would basically be async counterparts to `File.ReadLines`, so they would mirror what `File.ReadAllLinesAsync` does, but read line by line lazily. These would fill the remaining gap for async reading done by the `File` type, mentioned in https://github.com/dotnet/corefx/issues/11220.

**Comment by GrabYourPitchforks** [comment]
Is the `CancellationToken` parameter at the end correct? I thought the pattern was that the `IAsyncEnumerable<T>`-returning component didn't take a `CancellationToken` and that a token was instead passed via an extension method.

**Comment by khellang** [comment]
> Is the `CancellationToken` parameter at the end correct? I thought the pattern was that the `IAsyncEnumerable<T>`-returning component didn't take a `CancellationToken` and that a token was instead passed via an extension method.

I think so, yes. I'm assuming the `CancellationToken` parameter would be annotated with an `EnumeratorCancellationAttribute` and used inside the actual implementation, like this:

https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.Threading.Channels/src/System/Threading/Channels/ChannelReader.netcoreapp.cs#L19-L28

This was the only similar API I could find in code base:

https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.Threading.Channels/ref/System.Threading.Channels.netcoreapp.cs#L16

**Comment by khellang** [comment]
Any hint on when this will be up for review? 👼

**Comment by GrabYourPitchforks** [comment]
We usually get to APIs in chronological order (by date opened) unless they're marked as blocking.

**Comment by terrajobst** [comment]
* Looks good as proposed
* The `[EnumeratorCancellation]` attribute might be needed in the implementation if we use `yield return`, but it doesn't have to be part of the ref.

```C#
namespace System.IO
{
    public static partial class File
    {
        public static IAsyncEnumerable<string> ReadLinesAsync(string path, CancellationToken cancellationToken = default);
        public static IAsyncEnumerable<string> ReadLinesAsync(string path, System.Text.Encoding encoding, CancellationToken cancellationToken = default);
    }
}

**Comment by carlossanlop** [comment]
@khellang will you get a chance to work on this in the next week or two? Otherwise, I should move it to Future.

**Comment by khellang** [comment]
@carlossanlop I've already done most of the work, but I won't be able to polish it up and submit a PR in the coming weeks as I'm on vacation 😢

**Comment by carlossanlop** [comment]
No worries. Let's move it to Future then.

**Comment by vcsjones** [comment]
@eiriktsarpalis There was a PR open for this that didn't get merged, I think this API is blocked on https://github.com/dotnet/runtime/issues/20824 if I understood the PR's comments correctly.

**Comment by huoyaoyuan** [comment]
@khellang Now #20824 is unblocked. Do you want to work on this again?

**Comment by khellang** [comment]
Thanks for the heads up, @huoyaoyuan! I'm pretty swamped with Other Stuff right now, so I can't commit to anything. I can try to breathe life back into #43108 when I have time, but if someone beats me to it, I won't mind 😅

**Comment by AraHaan** [comment]
Wouldn't it be worth making ReadAllLinesAsync then call into this new API then to optimize them a bit as well?

**Comment by lateapexearlyspeed** [comment]
Hi, I would try to learn and fix it, not sure if can assign to me, here is my draft [PR](https://github.com/dotnet/runtime/pull/66492) which will be marked as ready, thanks.

**Comment by lateapexearlyspeed** [comment]
Hi, this [PR](https://github.com/dotnet/runtime/pull/66492) is marked as ready for review now, thanks.
